File: clients/fsharp/Model/Quest.fs
#nowarn "0058"

namespace Komus24
namespace Komus24.Model

open Komus24

/// A delivery quest
type Quest = {
    /// Cell where to pick delivery from
    PickupCell: Model.Vec2Int;
    /// Cell to drop the delivery at
    DropCell: Model.Vec2Int;
    /// Score for completing the quest
    Score: int64;
} with

    /// Write Quest to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        this.PickupCell.writeTo writer
        this.DropCell.writeTo writer
        writer.Write this.Score
        ()

    /// Read Quest from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        PickupCell = Model.Vec2Int.readFrom reader;
        DropCell = Model.Vec2Int.readFrom reader;
        Score = reader.ReadInt64()
    }

File: clients/csharp/Model/Quest.cs
namespace Komus24.Model
{
    /// <summary>
    /// TODO - Document
    /// </summary>
    public struct Quest
    {
        /// <summary>
        /// TODO - Document
        /// </summary>
        public Komus24.Model.Vec2Int PickupCell { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public Komus24.Model.Vec2Int DropCell { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public long Score { get; set; }
    
        public Quest(Komus24.Model.Vec2Int pickupCell, Komus24.Model.Vec2Int dropCell, long score)
        {
            this.PickupCell = pickupCell;
            this.DropCell = dropCell;
            this.Score = score;
        }
    
        /// <summary> Read Quest from reader </summary>
        public static Quest ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new Quest();
            result.PickupCell = Komus24.Model.Vec2Int.ReadFrom(reader);
            result.DropCell = Komus24.Model.Vec2Int.ReadFrom(reader);
            result.Score = reader.ReadInt64();
            return result;
        }
    
        /// <summary> Write Quest to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            PickupCell.WriteTo(writer);
            DropCell.WriteTo(writer);
            writer.Write(Score);
        }
    
        /// <summary> Get string representation of Quest </summary>
        public override string ToString() {
            string stringResult = "Quest { ";
            stringResult += "PickupCell: ";
            stringResult += PickupCell.ToString();
            stringResult += ", ";
            stringResult += "DropCell: ";
            stringResult += DropCell.ToString();
            stringResult += ", ";
            stringResult += "Score: ";
            stringResult += Score.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}